astropy/io/fits/hdu/compressed/compressed.py

```diff
@@ -687,292 +599,83 @@ def _summary(self):
-            self.data /= _scale
-            self.header["BSCALE"] = _scale
-        else:
-            for header in (self.header, self._header):
-                with suppress(KeyError):
-                    del header["BSCALE"]
-
-        if self.data.dtype.type != _type:
-            self.data = np.array(np.around(self.data), dtype=_type)  # 0.7.7.1
-
-        # Update the BITPIX Card to match the data
-        self._bitpix = DTYPE2BITPIX[self.data.dtype.name]
-        self._bzero = self.header.get("BZERO", 0)
-        self._bscale = self.header.get("BSCALE", 1)
-        # Update BITPIX for the image header specifically
-        # TODO: Make this more clear by using self._image_header, but only once
-        # this has been fixed so that the _image_header attribute is guaranteed
-        # to be valid
-        self.header["BITPIX"] = self._bitpix
-
-        # Update the table header to match the scaled data
-        self._update_header_data(self.header)
-
-        # Since the image has been manually scaled, the current
-        # bitpix/bzero/bscale now serve as the 'original' scaling of the image,
-        # as though the original image has been completely replaced
-        self._orig_bitpix = self._bitpix
-        self._orig_bzero = self._bzero
-        self._orig_bscale = self._bscale
+        self._add_data_to_bintable(self._tmp_bintable)
 
-    def _prewriteto(self, checksum=False, inplace=False):
-        if self._scale_back:
-            self.scale(BITPIX2DTYPE[self._orig_bitpix])
+        # If a bintable already exists internally we should update that instead
+        # of using a whole new BinTableHDU so that mode='update' works.
 
-        if self._has_data:
-            self._update_compressed_data()
-
-            # Use methods in the superclass to update the header with
-            # scale/checksum keywords based on the data type of the image data
-            self._update_pseudo_int_scale_keywords()
-
-            # Shove the image header and data into a new ImageHDU and use that
-            # to compute the image checksum
-            image_hdu = ImageHDU(data=self.data, header=self.header)
-            image_hdu._update_checksum(checksum)
-            if "CHECKSUM" in image_hdu.header:
-                # This will also pass through to the ZHECKSUM keyword and
-                # ZDATASUM keyword
-                self._image_header.set(
-                    "CHECKSUM",
-                    image_hdu.header["CHECKSUM"],
-                    image_hdu.header.comments["CHECKSUM"],
-                )
-            if "DATASUM" in image_hdu.header:
-                self._image_header.set(
-                    "DATASUM",
-                    image_hdu.header["DATASUM"],
-                    image_hdu.header.comments["DATASUM"],
-                )
-            # Store a temporary backup of self.data in a different attribute;
-            # see below
-            self._imagedata = self.data
+        if self._bintable is not None:
```

Comment:
This is a philosophical/deliberate choice - if a ``CompImageHDU`` wasn't backed by a _bintable prior to writing, I was going with the idea that it should also not be after writing, i.e. writing the ``CompImageHDU`` should not have any side effects on ``CompImageHDU``. Does this make sense or do you think after writing to a file the ``CompImageHDU`` should switch to being backed by a bintable? (this could result in changes to ``.data`` if there is any quantization).

Reply:
Yes we need to, at least in update mode, as if the data isn't changed, we want to write the bintable out as-is without recompressing it. If we dropped ``_bintable`` when the data is decompressed we'd also lose the ability for the user to use the ``compressed_data`` property. I would vote to keep it as-is for now and to see in a follow-up whether there are any benefits to dropping ``_bintable`` once ``.data`` is accessed under certain circumstances.